Add Reminder plugin

themegaphoenix opened the pull request:
This will allow the user to set reminder
Only tested in windows 10! Needs testing in Mac and Windows 7/8
Icon taken from: https://www.iconfinder.com/icons/285670/calendar_icon
ParseReminder: allows to parse expression in to reminders
Node-Schedule: allows to schedule events
Notifications: Default Electron notifications plugin

The current messages it can parse are:
- remind me to send invites at 3pm tomorrow
- remind me to wish Jamie happy birthday on 8/28
- remind me in 10 minutes to change the laundry
- remind me that the oil needs changed on Oct 4
- remind me next Friday to water the plants

Issue: #366

oliverschwendener commented:
Thanks for your contribution! I'll have a look and leave some comments.

oliverschwendener commented on package.json:
```diff
@@ -7,6 +7,9 @@
     "type": "git",
     "url": "https://github.com/oliverschwendener/ueli"
   },
+  "build": {
+    "appId": "ueli.com"
+  },
```
Why is this needed?

oliverschwendener commented on package.json:
```diff
@@ -65,7 +68,10 @@
     "bulma": "^0.8.0",
-    "bulma-extensions": "^6.2.7"
+    "bulma-extensions": "^6.2.7",
+    "node-schedule": "^1.3.2",
+    "parse-reminder": "^1.2.0"
```
Do we really need these 3rd party packages? Can't this be implemented without any extra dependencies?

themegaphoenix replied:
ParseReminder: allows to parse expression in to reminders - it simplifies development as it allows to break down the reminder expression to 'what', 'when' and 'where' which can then be used for the notifications
Node-Schedule: allows to schedule events - you can use the date parsereminder date to schedule notifications

stefnotch replied:
It should be quite possibly to replace parse reminder with a built-in implementation, as its code is rather simple. Replacing chrono-js, which parse-reminder depends on seems to be a fair bit harder. https://github.com/bkeepers/parse-reminder/blob/master/index.js

Replacing node-schedule with a custom made solution should be rather simple, because it doesn't actually schedule a cronjob or anything. Instead, "scheduled jobs will only fire as long as your script is running".

themegaphoenix replied:
@stefnotch i will look in replacing parse-reminder with a custom one based on it - but having a look at the chrono-js it is beyond my capabilities right now
regarding the node-schedule i think a timeout would work? it would have the same capabilities. How would we go about notification when the code is not running i.e. restart?

stefnotch replied:
I suppose it would be acceptable to assume that ueli is always running and only show reminders while ueli is running. Regarding an ueli restart, I think you pretty much have to write the reminders to a file or somehow persist them. But I'm not sure, so I'm pinging @oliverschwendener

themegaphoenix replied:
@stefnotch yeah you are right about a file/sqlite database for the persist. and load them everytime the reminder load. 
Would a calendar integration be better for this?

themegaphoenix replied:
@stefnotch @oliverschwendener i have implemented my own parse reminder. However for the scheduler i am unsure how to do it. Settimout max value is around 24 days. Would this be sufficient?

oliverschwendener commented on src/common/config/reminder-options.ts:
```diff
@@ -0,0 +1,9 @@
+export interface ReminderOptions {
+    isEnabled: boolean;
+    prefix:string;
```
Please pay attention to formatting. Here is a missing space.

oliverschwendener commented on src/common/icon/default-icons.ts:
```diff
@@ -1,6 +1,6 @@
-import { Icon } from "./icon";
+import {Icon} from "./icon";
 
-import { IconType } from "./icon-type";
+import {IconType} from "./icon-type";
```
These changes are unnecessary.

themegaphoenix replied:
This changes were made by WebStorm when formatting the file.
Should I change it back or keep it?

oliverschwendener replied:
Change it back.

oliverschwendener commented on src/common/translation/german-translation-set.ts:
```diff
@@ -345,4 +345,8 @@ export const germanTranslationSet: TranslationSet = {
+
+    reminder: "Erinnerung",
+    reminderPrefix: "Präfix",
+    reminderSettingsDescription: "Mit diesem Plugin können Sie Erinnerungen festlegen\n",
```
Why is `\n` neede here?

themegaphoenix replied:
Good catch!
It seems that when i copied and pasted from Google Translate it brought it with it.
I will fix it

oliverschwendener commented on src/common/translation/russian-translation-set.ts:
```diff
@@ -345,4 +345,8 @@ export const russianTranslationSet: TranslationSet = {
+
+    reminder: "напоминание",
+    reminderPrefix: "Префикс",
+    reminderSettingsDescription: "Этот плагин позволяет устанавливать напоминания\n",
```
Why is `\n` neede here?

themegaphoenix replied:
Good catch!
It seems that when i copied and pasted from Google Translate it brought it with it.
I will fix it

oliverschwendener commented on src/main/main.ts:
```diff
@@ -15,7 +15,7 @@ import { getProductionSearchEngine } from "./production/production-search-engine
 import { defaultGeneralOptions } from "../common/config/general-options";
 import { getErrorSearchResultItem } from "../common/error-search-result-item";
-import { FileHelpers } from "./../common/helpers/file-helpers";
+import { FileHelpers } from "../common/helpers/file-helpers";
```
This change is not related to the actual purpose of this PR.

oliverschwendener commented on src/main/main.ts:
```diff
@@ -460,7 +460,7 @@ function mainWindowNeedsToBeTransparent(userConfigOptions: UserConfigOptions): b
     }
 
-    return userConfigOptions.appearanceOptions.allowTransparentBackground === true;
+    return userConfigOptions.appearanceOptions.allowTransparentBackground;
```
This change is not related to the actual purpose of this PR.

oliverschwendener commented on src/main/main.ts:
```diff
@@ -470,7 +470,7 @@ function getMainWindowBackgroundColor(userConfigOptions: UserConfigOptions): str
     }
 
-    return userConfigOptions.appearanceOptions.allowTransparentBackground === true
+    return userConfigOptions.appearanceOptions.allowTransparentBackground
```
This change is not related to the actual purpose of this PR.

oliverschwendener commented on src/main/main.ts:
```diff
@@ -757,6 +757,7 @@ app.on("ready", () => {
         quitApp();
     }
 });
+app.setAppUserModelId("ueli.com");
```
Can you explain why this is needed?

themegaphoenix replied:
the reason that i added this is line is because electron notification doesn't show otherwise in windows:
https://www.electronjs.org/docs/tutorial/notifications#windows

themegaphoenix replied:
this the notification with the line:
![image](https://user-images.githubusercontent.com/9766462/81399328-bd5b1500-9122-11ea-9cf2-5edbba4d1d49.png)
this is without:
![image](https://user-images.githubusercontent.com/9766462/81399462-f4c9c180-9122-11ea-949c-71a0272e3f19.png)
do you want to change it to something else or remove the line entirely?

oliverschwendener commented on src/main/plugins/reminder-plugin/reminder-plugin.ts:
```diff
@@ -0,0 +1,94 @@
+import {ReminderOptions} from "../../../common/config/reminder-options";
+import {defaultReminderIcon} from "../../../common/icon/default-icons";
+import {join} from "path";
+// @ts-ignore
```
Can you explain why you added the `@ts-ignore` here?

themegaphoenix replied:
withouth that line i get:
`TS7016: Could not find a declaration file for module 'parse-reminder'. 'C:\Users\USERNAME\OneDrive\Desktop\ueli\node_modu
les\parse-reminder\index.js' implicitly has an 'any' type.
`
Do you know how to fix this? I tried googling but I seem unable to get a solution.
It is my first time using Typescript so i dont know the inside out this errors

Changing to `import {parseReminder} from "parse-reminder";` still gives me the same error

oliverschwendener commented on src/common/icon/default-icons.ts:
```diff
@@ -95,3 +95,8 @@ export const defaultBookmarkIcon: Icon = {
 
 export const defaultControlPanelIcon = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGAAAABgCAYAAADimHc4AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAABkwSURBVHhe7VwJsF1Vld3vv/HPQ/6Q/0MCnwBBAhEQurWRVlFoB6S11QKV0i6HFpVod9mtJVS1FqXViFV0i1oqVou29qC0Q6SRdqARlEKiaYwMkSEhJD/Jn+f/5qH32nfv88597/2QALFI9VvFPnuddc499969zz13+C9QE0000UQTTTTRRBNNNNFEE0000UQTTTTRRBNNHHtEhi7+SHv7QPc1LR1dV ...
+
+export const defaultReminderIcon :Icon= {
```
Here is a missing whitespace. Please format the code like this:

``` typescript
const defaultReminderIcon: Icon = { ... }
```

oliverschwendener commented on src/common/icon/default-icons.ts:
```diff
@@ -95,3 +95,8 @@ export const defaultBookmarkIcon: Icon = {
 export const defaultControlPanelIcon = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGAAAABgCAYAAADimHc4AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAABkwSURBVHhe7VwJsF1Vld3vv/HPQ/6Q/0MCnwBBAhEQurWRVlFoB6S11QKV0i6HFpVod9mtJVS1FqXViFV0i1oqVou29qC0Q6SRdqARlEKiaYwMkSEhJD/Jn+f/5qH32nfv88597/2QALFI9VvFPnuddc499969zz13+C9QE0000UQTTTTRRBNNNNFEE0000UQTTTTRRBNNHHtEhi7+SHv7QPc1LR1dV ...
+
+export const defaultReminderIcon :Icon= {
+    parameter:`<svg version="1.1" xmlns="http://www.w3.org/2000/svg"  xmlns:cc="http://creativecommons.org/ns#" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"><g transform="translate(0 -1028.4)"><path d="m5 1032.4c-1.1046 0-2 0.9-2 2v14c0 1.1 0.8954 2 2 2h6 2 6c1.105 0 2-0.9 2-2v-14c0-1.1-0.895-2-2-2h-6-2-6z" fill="#bdc3c7"/><path d="m5 3c-1.10 ...
```
Here is a missing whitespace

oliverschwendener commented:
Can you please target the `dev` branch?

theRealPadster commented:
I have successfully built and ran it under macOS Catalina. 

It detected my reminder query (though it seems to only support "min/minutes/sec/seconds" and not "30 s/30 m". 

The first time a reminder fired off, it asked if I wanted to allow Electron to display reminders, which I granted. I then had to set another reminder, as the first one didn't show up.
![image](https://user-images.githubusercontent.com/1565883/81616642-dcbba180-93b1-11ea-9220-9f3248e779fa.png)

When the second one showed up, it had the Electron icon, not UELI. There is also no close button, like some notifications have, but it's not a huge deal. 
![image](https://user-images.githubusercontent.com/1565883/81616653-e0e7bf00-93b1-11ea-827f-48979628fe3c.png)

This is the Lacona reminder notification, for reference. The snoozing isn't a big deal, but it is something that I noticed is present. 
![image](https://user-images.githubusercontent.com/1565883/81616620-d3323980-93b1-11ea-823e-dbb020b0809a.png)

themegaphoenix commented:
> I have successfully built and ran it under macOS Catalina.
> 
> It detected my reminder query (though it seems to only support "min/minutes/sec/seconds" and not "30 s/30 m".
> 
> The first time a reminder fired off, it asked if I wanted to allow Electron to display reminders, which I granted. I then had to set another reminder, as the first one didn't show up.
> ![image](https://user-images.githubusercontent.com/1565883/81616642-dcbba180-93b1-11ea-9220-9f3248e779fa.png)
> 
> When the second one showed up, it had the Electron icon, not UELI. There is also no close button, like some notifications have, but it's not a huge deal.
> ![image](https://user-images.githubusercontent.com/1565883/81616653-e0e7bf00-93b1-11ea-827f-48979628fe3c.png)
> 
> This is the Lacona reminder notification, for reference. The snoozing isn't a big deal, but it is something that I noticed is present.
> ![image](https://user-images.githubusercontent.com/1565883/81616620-d3323980-93b1-11ea-823e-dbb020b0809a.png)

regarding the authorisation, i think when enabling a reminder plugin should also show up a new notification so that the user can authorise it? it would be a dirty trick 
for the icon i think because of the relative icon the path is causing problems.
To help me diagnose can you add this line after the `exectute` function starts line 70 (before the promise)
` console.log(join(__dirname,'../img/icons/win/icon-transparent.ico'));`
this would tell me what directory the file is running from in the mac and thus issue a fix

regarding the extra buttons i think it could be added:
https://www.electronjs.org/docs/api/structures/notification-action
not exactly sure about it though

theRealPadster commented:
> for the icon i think because of the relative icon the path is causing problems.
> To help me diagnose can you add this line after the `exectute` function starts line 70 (before the promise)
> ` console.log(join(__dirname,'../img/icons/win/icon-transparent.ico'));`
> this would tell me what directory the file is running from in the mac and thus issue a fix
> regarding the extra buttons i think it could be added:
> https://www.electronjs.org/docs/api/structures/notification-action
> not exactly sure about it though



It logs: `/Users/(username)/Sources/ueli/img/icons/win/icon-transparent.ico`

I took a look at the button action thing, but I don't know anything about developing with Electron or Typescript, and it didn't make much sense to me, so I'm not sure how much help I can be there.

themegaphoenix commented:
> > for the icon i think because of the relative icon the path is causing problems.
> > To help me diagnose can you add this line after the `exectute` function starts line 70 (before the promise)
> > ` console.log(join(__dirname,'../img/icons/win/icon-transparent.ico'));`
> > this would tell me what directory the file is running from in the mac and thus issue a fix
> > regarding the extra buttons i think it could be added:
> > https://www.electronjs.org/docs/api/structures/notification-action
> > not exactly sure about it though
> 
> It logs: `/Users/(username)/Sources/ueli/img/icons/win/icon-transparent.ico`
> 
> I took a look at the button action thing, but I don't know anything about developing with Electron or Typescript, and it didn't make much sense to me, so I'm not sure how much help I can be there.

thank you for this! Does the path exists and is the icon there? If the icon is there I have no clue why it doesnt show up

theRealPadster commented:
The icon is there. I'm not sure if it's an issue with it being in .ico format or what, but it previews fine when I try and open the image.

themegaphoenix commented:
> The icon is there. I'm not sure if it's an issue with it being in .ico format or what, but it previews fine when I try and open the image.

i think this might be of interest:
https://stackoverflow.com/questions/37547559/cannot-change-electron-notification-icon-on-left

theRealPadster commented:
> > The icon is there. I'm not sure if it's an issue with it being in .ico format or what, but it previews fine when I try and open the image.
> 
> i think this might be of interest:
> https://stackoverflow.com/questions/37547559/cannot-change-electron-notification-icon-on-left
> refreshing the cache might help:
> https://stackoverflow.com/questions/11856766/osx-notification-center-icon


If I change the icon path to one of the mac/png ones, it shows up on the right side, but not the app icon. 
![image](https://user-images.githubusercontent.com/1565883/82473235-42391d80-9a97-11ea-9aa3-f3c853c8ada2.png)

The app still identifies itself as Electron for some reason, which would be why the icon is Electron. 
![image](https://user-images.githubusercontent.com/1565883/82473379-78769d00-9a97-11ea-8a0a-7c61c51345e6.png)

I googled electron app-bundle-id and found [this](https://github.com/electron-userland/electron-builder/issues/368),but it didn't seem to do anything when I added this to the package.json and rebuilt it. 
```
"build": {
    "appId": "org.test.testing"
  },
```

I'll have to do some more poking around later and see what I can dig up.

themegaphoenix commented:
> > > The icon is there. I'm not sure if it's an issue with it being in .ico format or what, but it previews fine when I try and open the image.
> > 
> > 
> > i think this might be of interest:
> > https://stackoverflow.com/questions/37547559/cannot-change-electron-notification-icon-on-left
> > refreshing the cache might help:
> > https://stackoverflow.com/questions/11856766/osx-notification-center-icon
> 
> If I change the icon path to one of the mac/png ones, it shows up on the right side, but not the app icon.
> ![image](https://user-images.githubusercontent.com/1565883/82473235-42391d80-9a97-11ea-9aa3-f3c853c8ada2.png)
> 
> The app still identifies itself as Electron for some reason, which would be why the icon is Electron.
> ![image](https://user-images.githubusercontent.com/1565883/82473379-78769d00-9a97-11ea-8a0a-7c61c51345e6.png)
> 
> I googled electron app-bundle-id and found [this](https://github.com/electron-userland/electron-builder/issues/368),but it didn't seem to do anything when I added this to the package.json and rebuilt it.
> 
> ```
> "build": {
>     "appId": "org.test.testing"
>   },
> ```
> 
> I'll have to do some more poking around later and see what I can dig up.

thank you very much.
I actually had those lines in package.json before in the earlier commits but i removed it because i added them for a previous dependency i used to make the notifications, but then i realised that electron already had it built in
if you can do this it would be helpful:
1. uninstall your current ueli installation
2. add the build line in package.json
3. run `yarn package` to create a installer
4. install the app

hopefully this will fix it

theRealPadster commented:
Ah, maybe it was because I was just running it from the command line and didn't actually build/install it. I did what you said and it seems to work now. It shows up as UELI, not Electron. The "icon" specified is still the one on the left, so we probably don't need it if it'll automatically use the UELI icon. 

![image](https://user-images.githubusercontent.com/1565883/82478682-41a48500-9a9f-11ea-9a1d-a7333e8d3dad.png)

themegaphoenix commented:
> Ah, maybe it was because I was just running it from the command line and didn't actually build/install it. I did what you said and it seems to work now. It shows up as UELI, not Electron. The "icon" specified is still the one on the left, so we probably don't need it if it'll automatically use the UELI icon.
> 
> ![image](https://user-images.githubusercontent.com/1565883/82478682-41a48500-9a9f-11ea-9a1d-a7333e8d3dad.png)

good to hear that everything works!
i will change so to that it checks which OS it is running and add the icon if needed
maybe even extract it the notification in its own helper function
thank you very much

theRealPadster commented:
Were you able to make any progress @themegaphoenix ? I've been looking forward to switching over from Lacona ;P

themegaphoenix commented:
@theRealPadster unfortunately I have not had much time lately.
I extract the notfication in to its own helper function and checked for the OS to set the icon.
Please test it

theRealPadster commented:
> @theRealPadster unfortunately I have not had much time lately.
> I extract the notfication in to its own helper function and checked for the OS to set the icon.
> Please test it

Ah, that's fair. If there's anything I can do to help, let me know :)
I pulled from your reminder branch (66 commits) and ran a `yarn package`, and installed it. I have the plugin enabled but for some reason it's not showing the reminder action when I try and run it. Is it working for you in Windows? Maybe I did something wrong?

themegaphoenix commented:
> Any updates on this?

This branch is now really outdated, so I will have to cancel this PR. Unfortunately I won't be able to add the feature :(
